test/jdk/java/net/httpclient/http3/H3LogHandshakeErrors.java

```diff
@@ -0,0 +1,175 @@
+ * @library /test/lib /test/jdk/java/net/httpclient/lib
+ * @build jdk.test.lib.net.SimpleSSLContext
+ *        jdk.httpclient.test.lib.common.HttpServerAdapters
+ * @run testng/othervm
+ *              -Djdk.httpclient.HttpClient.log=errors
+ *              H3LogHandshakeErrors
+ */
+// -Djava.security.debug=all
+public class H3LogHandshakeErrors implements HttpServerAdapters {
+
+    private SSLContext sslContext;
+    private HttpTestServer h3Server;
+    private String requestURI;
+
+    @BeforeClass
+    public void beforeClass() throws Exception {
+        sslContext = new SimpleSSLContext().get();
+        if (sslContext == null) {
+            throw new AssertionError("Unexpected null sslContext");
+        }
+        // create an H3 only server
+        h3Server = HttpTestServer.create(HTTP_3_URI_ONLY, sslContext);
+        h3Server.addHandler((exchange) -> exchange.sendResponseHeaders(200, 0), "/hello");
+        h3Server.start();
+        System.out.println("Server started at " + h3Server.getAddress());
+        requestURI = "https://" + h3Server.serverAuthority() + "/hello";
+    }
+
+    @AfterClass
+    public void afterClass() throws Exception {
+        if (h3Server != null) {
+            System.out.println("Stopping server " + h3Server.getAddress());
+            h3Server.stop();
+        }
+    }
+
+    static String format(LogRecord record) {
+        String thrown = Optional.ofNullable(record.getThrown())
+                .map(t -> ": " + t).orElse("");
+        return "\n  \"%s: %s %s: %s%s\"".formatted(
+                record.getLevel(),
+                record.getSourceClassName(),
+                record.getSourceMethodName(),
+                record.getMessage(),
+                thrown);
+    }
+
+    /**
+     * Issues a GET HTTP3 requests and verifies that the
+     * expected exception is logged.
+     */
+    @Test
+    public void testErrorLogging() throws Exception {
+        // jdk.httpclient.HttpClient.log=errors must be enabled
+        // for this test
+        String logging = System.getProperty("jdk.httpclient.HttpClient.log", "");
+        var categories = Arrays.asList(logging.split(","));
+        assertTrue(categories.contains("errors"),
+                "'errors' not found in " + categories);
+
+        // create a client that doesn't have the server's
+        // certificate
+        final HttpClient client = newClientBuilderForH3()
+                .proxy(NO_PROXY)
+                .build();
+        final URI reqURI = new URI(requestURI);
+        final HttpRequest.Builder reqBuilder = HttpRequest.newBuilder(reqURI)
+                .version(HTTP_3);
+        Logger serverLogger = Logger.getLogger("jdk.httpclient.HttpClient");
+
+        CopyOnWriteArrayList<LogRecord> records = new CopyOnWriteArrayList<>();
+        Handler handler = new Handler() {
+            @Override
+            public void publish(LogRecord record) {
+                record.getSourceClassName();
```

Comment:
Is this line a leftover? Or does it intentionally call this method and not use the return value?

Reply:
It does intentionally call it to force the LogRecord to infer the caller at the right place. Inferring the caller has to happen in the handler to ensure the right caller will be found.